Updated view_user_cert to support all certificates.

Raza403 opened the pull request:
Previously, this was working only for some specific certificates.

Raza403 committed:
customcert user's has option to use new or old format. Unique code is being generated. This PR is made for #1 and #2

custom cert, updated the names of both code generation methods.

Added view_user_cert page, a logged in user can see certificate if having a valid token and ecard code. This is being used from view_tokens.php page.

customcert view certificate function is made. This is made for #212

Updated verify certificate URL. This PR is made for 221

Updated the verify URL for QR code.

Raza403 committed:
Updated view_user_cert to view all certs including CE certs. This PR is made for #231

mdjnelson commented:
Can you please look at failing CI pipeline please?

mdjnelson commented:
Please rebase on the latest MOODLE_404_STABLE branch as I have bumped the Moodle Plugin CI version.

mdjnelson commented:
This is still failing. Ill have a look at the code and fix the failures in another commit if all is good.

mdjnelson commented on classes/certificate.php:
```diff
@@ -552,49 +552,62 @@ public static function issue_certificate($certificateid, $userid) {
-     *
-     * @return string
-     */
-    public static function generate_code(): string {
```
Why is the return type being removed?

mdjnelson commented on classes/certificate.php:
```diff
@@ -552,49 +552,62 @@ public static function issue_certificate($certificateid, $userid) {
-            random_int(0, 9999)
-        );
+    // Digits with hyphens
+    private static function generate_code_digits_with_hyphens() {
```
Again, return type missing.

mdjnelson commented on db/upgrade.php:
```diff
@@ -298,21 +298,23 @@ function xmldb_customcert_upgrade($oldversion) {
     }
 
+    // Drop the unique index on 'code' and add a non-unique one.
     if ($oldversion < 2024042210) {
```
Why are we touching this upgrade step?

mdjnelson commented on element/qrcode/classes/element.php:
```diff
@@ -173,7 +172,7 @@ public function render($pdf, $preview, $user) {
             }
 
-            $qrcodeurl = new \moodle_url('/mod/customcert/verify_certificate.php', $urlparams);
+            $qrcodeurl = new \moodle_url('https://app.pacificmedicaltraining.com/verify', $urlparams);
```
You have hardcoded a URL here.

mdjnelson commented on settings.php:
```diff
@@ -85,10 +85,10 @@
     'customcert/codegenerationmethod',
     get_string('codegenerationmethod', 'customcert'),
     get_string('codegenerationmethod_desc', 'customcert'),
-    0, // Default option (0 = Upper/lower/digits random string method).
```
Why remove the '.'?

mdjnelson commented on settings.php:
```diff
@@ -85,10 +85,10 @@
     [
-        0 => get_string('codegenerationmethod_upperlowerdigits', 'customcert'), // Upper/lower/digits random string.
-        1 => get_string('codegenerationmethod_digitshyphens', 'customcert'), // Digits with hyphens numeric code.
+        0 => get_string('Upper/lower/digits', 'customcert'), // Upper/lower/digits random string
```
Why do we need different language strings?

mdjnelson commented:
Apologies. Initially I said I would address these but there are a few uncertainties about this patch. I would like this cleaned up so I can see a clear solution (right now it's touching code all over the place and im not sure why) at what is trying to be achieved. Please let me know when you are ready for me to re-review. Thanks.

Raza403 committed:
removed some errors in the code

Raza403 committed:
removed some errors in the code

fulldecent committed:
Update upgrade.php

As per https://github.com/mdjnelson/moodle-mod_customcert/pull/681#pullrequestreview-3000773525

Raza403 committed:
customcert fixed errors.

Raza403 committed:
Fixed errors.

Raza403 committed:
Fixed errors

Raza403 committed:
Fixed errors

Raza403 committed:
fixed errors

Raza403 committed:
custom_cert fixed the phpdoc errors.

Raza403 committed:
Allowed guest access.

Raza403 committed:
Fixed an error in lib.php, to prevent direct access

Raza403 committed:
custom cert fixed an error

fulldecent commented on view_user_cert.php:
```diff
@@ -0,0 +1,110 @@
+require_once($CFG->dirroot . '/mod/customcert/lib.php');
+
+// Allows guest access to course with ID 1.
+require_login(1, false);
```
Does this line cause this script to only work if the site is configured with "allow guest login" ?

Are there any other situations where this line of code would cause the certificate to not be displayed?
